src/audio/pcm_converter/pcm_converter_generic.c

```diff
@@ -431,3 +432,132 @@ const struct pcm_func_map pcm_func_map[] = {
 const size_t pcm_func_count = ARRAY_SIZE(pcm_func_map);
 
 #endif
+
+static int pcm_convert_s16_c16_to_s16_c32(const struct audio_stream *source,
+					  uint32_t ioffset, struct audio_stream *sink,
+					  uint32_t ooffset, uint32_t samples)
+{
+	uint32_t buff_frag = 0;
+	int16_t *src;
+	int32_t *dst;
+	uint32_t i;
+
+	for (i = 0; i < samples; i++) {
+		src = audio_stream_read_frag_s16(source, buff_frag + ioffset);
+		dst = audio_stream_write_frag_s32(sink, buff_frag + ooffset);
+		*dst = *src;
+		buff_frag++;
+	}
+
+	return samples;
+}
+
+static int pcm_convert_s16_c32_to_s16_c16(const struct audio_stream *source,
+					  uint32_t ioffset, struct audio_stream *sink,
+					  uint32_t ooffset, uint32_t samples)
+{
+	uint32_t buff_frag = 0;
+	int32_t *src;
+	int16_t *dst;
+	uint32_t i;
+
+	for (i = 0; i < samples; i++) {
+		src = audio_stream_read_frag_s32(source, buff_frag + ioffset);
+		dst = audio_stream_write_frag_s16(sink, buff_frag + ooffset);
+		*dst = *src & 0xffff;
+		buff_frag++;
+	}
+
+	return samples;
+}
+
+static int pcm_convert_s16_c32_to_s32_c32(const struct audio_stream *source,
+					  uint32_t ioffset, struct audio_stream *sink,
+					  uint32_t ooffset, uint32_t samples)
+{
+	uint32_t buff_frag = 0;
+	int32_t *src, *dst;
+	uint32_t i;
+
+	for (i = 0; i < samples; i++) {
+		src = audio_stream_read_frag_s32(source, buff_frag + ioffset);
+		dst = audio_stream_write_frag_s32(sink, buff_frag + ooffset);
+		*dst = (*src & 0xffff) << 16;
```

Comment:
The and operation is not necessary, leaving it out would make the code faster. But there was a discussion earlier that C standard does define overflow in the arithmetic shift. If it would be saturating type, that would fail. We know it's not saturating in gcc/xt-xcc and lot of code in SOF relies in that but to be safe with all compilers/platforms this may be needed.

Reply:
thanks, for this case high 16 should be zero. So let's remove and operation to for debugging hint.